### ISSUE TITLE
Integration test `test_format_avro_confluent` is flaky

### ISSUE THREAD

**Issue by KochetovNicolai** [ClickHouse/ClickHouse#67463]
Tests `test_format_avro_confluent/test.py::test_select`, `test_format_avro_confluent/test.py::test_select_auth`, `test_format_avro_confluent/test.py::test_select_auth_encoded`, `test_format_avro_confluent/test.py::test_select_auth_encoded_complex` fail in batch with

```
E               Exception: Command ['docker-compose', '--env-file', '/ClickHouse/tests/integration/test_format_avro_confluent/_instances_0/.env', '--project-name', 'roottestformatavroconfluent', '--file', '/ClickHouse/tests/integration/helpers/../../../tests/integration/compose/docker_compose_kafka.yml', '--verbose', 'up', '-d', '--renew-anon-volumes'] return non-zero code 1: compose.config.config.find: Using configuration files: /ClickHouse/tests/integration/helpers/../../../tests/integration/compose/docker_compose_kafka.yml
```

CI database

https://play.clickhouse.com/play?user=play#U0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIGNoZWNrX3N0YXR1cywgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnT0snCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwogICAgQU5EIHRlc3RfbmFtZSBsaWtlICcldGVzdF9mb3JtYXRfYXZyb19jb25mbHVlbnQlJwogICAgQU5EIGNoZWNrX3N0YXJ0X3RpbWUgPiBub3coKSAtIGludGVydmFsIDYgbW9udGgKT1JERVIgQlkgY2hlY2tfc3RhcnRfdGltZSBkZXNjLCBjaGVja19uYW1lLCB0ZXN0X25hbWU=

Reports

https://s3.amazonaws.com/clickhouse-test-reports/0/5e41c29a90e50dce7cabe5a6a260a5690721dd2e/integration_tests__asan__[1_4].html
https://s3.amazonaws.com/clickhouse-test-reports/0/dfcefab9ab906c9bb42556a12923db626632b8b0/integration_tests__tsan__[5_6].html
https://s3.amazonaws.com/clickhouse-test-reports/0/823afac5aeebc5b7194c9086b0bd1430b0dfa7dc/integration_tests__asan__old_analyzer__[5_6].html

**Comment by Avogar** [comment]
So, the problem here is in ports clashing:
```
E               ERROR: for schema-registry  Cannot start service schema-registry: driver failed programming external connectivity on endpoint roottestformatavroconfluent_schema-registry_1 (26724227c33237ea6f004473fe99de25aeaa2e613e7f8cd96cf2e5fd102ac0e3): Bind for 0.0.0.0:56397 failed: port is already allocated
```

And in all cases we just allocated the same port to kafka and to schema registry. Ports allocation is hapenning using this function:
https://github.com/ClickHouse/ClickHouse/blob/58998d70694dc07048b16d67c041eef91473afeb/tests/integration/helpers/cluster.py#L131-L135

And it seems like it can produce the same port several times. There were a few attempts to fix it:
https://github.com/ClickHouse/ClickHouse/pull/55055 - rebind port in case of clashing
And
https://github.com/ClickHouse/ClickHouse/pull/53266 - do not specify host port while exposing ports from container

First, I tried the second approach, but after long time of debugging understood that it's impossible: in kafka we use `KAFKA_ADVERTISED_LISTENERS` that always requires both inside and outside ports.

The first approach with rebinding pors seems to be too much, from what I see, the problem is in allocating the same ports for different services, not allocating port that is already in use.

So, I think we can just add global list of all ports returned in get_free_port and check if we already returned this port to avoid clashing
